### ISSUE TITLE
label2rgb has an undocumented implicit assumption that labels are consecutive non-negative integers

### ISSUE THREAD

**Issue by rfinck** [scikit-image/scikit-image#2225]
## Description

An implicit assumption that the labels provided to label2rgb are consecutive non-negative integers can produce unexpected behavior. Two examples below are negative labels resulting in failure to assign background color correctly, and missing labels combined with color cycling.

I can provide a fix unless there is a desire to insist on consecutive non-negative integer labels, in which case this should raise a helpful exception.
## Way to reproduce

``` python
labels = np.array([
    [0, 1],
    [2, 0]
])
fig, ax = plt.subplots(2, 2)
ax[0, 0].imshow(labels)
ax[0, 0].set_title('Nonnegative Labels')
ax[0, 1].imshow(skimage.color.label2rgb(labels, bg_label=0))
ax[0, 1].set_title('RGB with Black Background')
ax[1, 0].imshow(-labels)
ax[1, 0].set_title('Negative Labels')
ax[1, 1].imshow(skimage.color.label2rgb(-labels, bg_label=0))
ax[1, 1].set_title('RGB with Buggy Blue Background')
```

<img width="522" alt="screen shot 2016-07-29 at 10 47 48 pm" src="https://cloud.githubusercontent.com/assets/1189568/17268399/670f4bea-55de-11e6-8834-c01a5b47338f.png">

``` python
labels = np.array([
    [0, 2],
    [4, 0]
])
fig, ax = plt.subplots(1, 2)
ax[0].imshow(labels)
ax[0].set_title('Grayscale Labels')
ax[1].imshow(skimage.color.label2rgb(labels, colors=[(1, 0, 0), (0, 0, 1)]))
ax[0, 1].set_title('RGB')
```

<img width="485" alt="screen shot 2016-07-29 at 11 08 58 pm" src="https://cloud.githubusercontent.com/assets/1189568/17268521/5af7a354-55e1-11e6-94c2-39add3706c8a.png">

scikit-image=0.12.3
Python 2.7.11

**Comment by stefanv** [comment]
@rfinck Thanks for reporting; I think your use case is legitimate.  A fix + test would be welcome.

**Comment by rfinck** [comment]
After looking at the source more carefully I now realize that my description was slightly off -- there is an attempt to deal with negative labels but it is not effective. It might be a little while until I get around to it but I'll submit a PR.

**Comment by tseclaudia** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [label2rgb has an undocumented implicit assumption that labels are consecutive non-negative integers](https://trello.com/c/rxz8nDmZ/2-label2rgb-has-an-undocumented-implicit-assumption-that-labels-are-consecutive-non-negative-integers)

**Comment by ahojnnes** [comment]
Closed by #2370. Thanks for the bug report.
